decisions/adr-0007-vulcan-workflows.md

```diff
@@ -0,0 +1,60 @@
+
+### Using Snakemake for workflows
+
+The format of a Vulcan workflow is a Snakemake workflow. This may include steps in python, R, etc. Workflows may define containers via a Snakefile and conda yaml. Vulcan will keep track of containers and re-use them across workflows.
+
+Inputs to the workflow can come from two places: 1) the data library, via https get, and 2) the cache/ folder associated with a Vulcan figure. Similarly, outputs can be: 1) writes to the data library (Magma or Metis) and 2) writes to the cache/ directory of a given folder
+
+Workflows will be executed by running Snakemake in the appropriate environment. Each job will contain a Snakefile, a config.yml, Vulcan inputs and an Etna token, and will generate output which will be copied into Vulcan's cache when the workflow completes or halts. Here we remain agnostic about how jobs will be dispatched.
+
+### UI Queries
+
+Currently the Vulcan orchestrator has special hooks for "ui query" steps. Since we cannot expect the Snakemake orchestator to understand this logic, we will instead separate these out into a new set of Vulcan endpoints, which allow a workflow to set a requirement for a UI query associated with a particular step in the workflow. To make use of this interface, a step in a Snakemake workflow which expects to output a UI query JSON (i.e., the JSON response from the user) will use a vulcan client interface to set a query requirement and exit, halting the workflow due to the unsatisfied output requirement. The Vulcan UI will then be able to determine an unresolved query and will show it to the user in the UI, making use of the existing query interfaces to resolve the query. This results in the response being written to the appropriate output location, and the Snakemake workflow resumes with the requirement satisfied.
+
+### Using Vulcan via the Command-Line
+
+Although it may eventually be attractive to run a given workflow in the Vulcan UI, the development loop in the UI is difficult. Files cannot be inspected directly, and other utilities, REPLs, etc, may not be accessible. Although the use of Snakemake workflows makes it easier to transfer Vulcan workflows to other environments and run them there, many dependencies (containers and inputs) may be more difficult to achieve than a git clone. To allow easier development, Vulcan will include a command-line client, which will run the workflow's Snakefile from the command-line in a given environment, allowing users to use Vulcan containers and attach inputs in local folders. Similarly the command-line client will allow a user to resolve a UI query step in the shell by posting the appropriate output JSON to the correct location.
```

Comment:
>Similarly the command-line client will allow a user to resolve a UI query step in the shell by posting the appropriate output JSON to the correct location

These JSONs can sometimes be VERY hard to create from scratch, and creation from scratch obfuscates all gotcha-catches and restrictions that an actual UI might have built in.  See below for an example.  **I think a browser-based (or something else that can render react) tool for UI input generation might be required.**  This could still be launched from the command line, and inherit its inputs by default.

<details>
 <summary>E.g. plot setups can be VERY large and complex:</summary>

AND, to name one of many quirks that the VizUI system handles, the `"plot_type": "dittoFreqPlot"` selection gets used by the VizUI system to determine what all the other fields can be because each plot types uses a different subset of the possible fields

```
plot_setup/plot_setup: {
  "plot_type": "dittoFreqPlot",
  "var": "wsnn.harmony.int_res.0.1",
  "group_by": "arm",
  "color_by": "make",
  "sample_by": "CoLabs_patient",
  "plots": [
    "jitter",
    "boxplot"
  ],
  "scale_by": "fraction",
  "plot_title": "make",
  "plot_subtitle": "make",
  "legend_title": "make",
  "xlab": "make",
  "ylab": "make",
  "group_order": "make",
  "vars_use": [],
  "cells_use": {},
  "vlnplot_lineweight": 1,
  "vlnplot_width": 1,
  "vlnplot_scaling": "area",
  "boxplot_width": 0.2,
  "boxplot_color": "black",
  "boxplot_fill": true,
  "boxplot_lineweight": 1,
  "jitter_size": 1,
  "jitter_width": 0.2,
  "jitter_color": "black",
  "ridgeplot_lineweight": 1,
  "legend_show": true,
  "split_adjust_free_y": true
}
```

</details>

Reply:
Fair - I thought that the command-line client could provide some help here, like an empty JSON template you can fill out. We can also work on simplifying/clarifying this format so it is easier to generate.